### DISCUSSION TITLE
How hard would it be to add quad SPI support for esp32-s3?

### DISCUSSION THREAD

**Discussion by echo-lalia** [micropython/micropython#14333] [ESP32]
Hey y'all! I just bought an [esp32 display module](https://github.com/profi-max/JC4827W543_4.3inch_ESP32S3_board) for pretty cheap. It didn't have a ton of available documentation, but my plans for the display are fairly simple, and so I figured I'd be able to get it running with a bit of work. 

From the examples shared in the repo I linked above, (and confirmed by the pinout) I noticed that the display is connected to the device using QSPI
```
Arduino_DataBus *bus = new Arduino_ESP32QSPI(
    45 /* cs */, 47 /* sck */, 21 /* d0 */, 48 /* d1 */, 40 /* d2 */, 39 /* d3 */);
```

Taking a look at the datasheet for the display driver, it looks like the display supports both "normal" SPI, and QSPI, however, it looks like you need to choose the SPI mode by setting a few pins on the driver chip, and I'm assuming those are set by some physical connections or something, not in software. 

<br/>

Anyways, I'd really like to program this device in MicroPython, if possible. However, I'm pretty sure there's no support for QSPI in the ESP32 port of MicroPython, and I'm struggling to understand what exactly needs to be done to enable this support. 

AFAIK the ESP-IDF already has the methods needed, and so I'm pretty sure it would come down to making a wrapper for those methods to expose them to MicroPython. However, I'm struggling to understand exactly what that would look like. 

I'm fairly new to working with embedded C, and my understanding is very limited on that side. I haven't been able to find any pull request or experimental implementations for QSPI in MicroPython, so I think I'd be working from scratch, kinda. 

So, I guess I'm asking:   
how big of a project is that going to be? My purpose for the display is a gift idea, so I have like a month to finish my work on it. Am I in over my head?

Does anyone have any resources that can be shared to help push me in the right direction? 

Thanks in advance 😊

[Related issue](https://github.com/micropython/micropython/issues/12124)

**Comment by echo-lalia** [comment]
Oh, another interesting note on this devboard:

It uses an Esp32-s3 WROOM with 4mb flash and 8MB Octal SPI RAM (or at least I *think* it's octal; I couldn't use the RAM without enabling it) 

To use the device in MicroPython, I already had to build a custom binary with both 4mb flash and octal-spiram support. I found this easy as I already have the build tools set up for making MicroPython. However, it did require making some changes to the board def :)

**Comment by xyzzy42** [comment]
I've added support for configuring the QSPI pins on ESP32.  It wasn't that hard.  But you would need to extend the `SPI.write()` and `SPI.read()` interface for writing data to also support configuring quad mode, which I haven't done.  A problem is the Micropython API for SPI has no concept of mode.

The display driver chip used in that display also support normal 3 and 4 pin SPI. You don't need to use QSPI.  It just makes it faster.

**Reply by echo-lalia** [reply]
Thanks for your input!

>The display driver chip used in that display also support normal 3 and 4 pin SPI. You don't need to use QSPI. It just makes it faster.

I'm pretty sure at this point that it would require surgery to enable this mode.    
The display isn't even hooked up to the board with an "rst" or "dc" pin, and it has not responded in any way to any configuration I tried with regular SPI :' )

That said, I actually did get it working (though it's a little janky) using [this](https://github.com/nspsck/RM67162_Micropython_QSPI/tree/main) MicroPython C module, which is designed for a completely different display (LILYGO T-DisplayS3-AMOLED). Just had to add a custom initialization function (and I'm also wrapping it in an additional MicroPython FrameBuffer).

**Reply by trentp-igor** [reply]
You should be able to drive it without using quad mode.  It's the same sub-set of pins on the LCD chip for any mode that are connected to the host.  You just need to configure the host to use the right ones.

It's also interesting that that driver would work, without 'dc'
```
{ MP_QSTR_dc,               MP_ARG_OBJ | MP_ARG_KW_ONLY | MP_ARG_REQUIRED        },
```
Looks like it requires it.

**Reply by echo-lalia** [reply]
Yeah like I said my solution is kinda janky; I'm just passing an unused pin as DC, and it works fine.

**Reply by trentp-igor** [reply]
Funny, while the driver requires a DC pin be specific, if you look through the code, it's clearly never used anywhere.  The LCD controller doesn't require DC, it just takes extra bits to use protocol version without it.

**Reply by picpic020960** [reply]
> Thanks for your input!
> 
> > The display driver chip used in that display also support normal 3 and 4 pin SPI. You don't need to use QSPI. It just makes it faster.
> 
> I'm pretty sure at this point that it would require surgery to enable this mode. The display isn't even hooked up to the board with an "rst" or "dc" pin, and it has not responded in any way to any configuration I tried with regular SPI :' )
> 
> That said, I actually did get it working (though it's a little janky) using [this](https://github.com/nspsck/RM67162_Micropython_QSPI/tree/main) MicroPython C module, which is designed for a completely different display (LILYGO T-DisplayS3-AMOLED). Just had to add a custom initialization function (and I'm also wrapping it in an additional MicroPython FrameBuffer).

Hello
Can you explain your work to help beginners.
Thank's

**Reply by echo-lalia** [reply]
> Hello Can you explain your work to help beginners. Thank's

My solution wasn't very elegant. Take a look at my repo [here](https://github.com/echo-lalia/beach-portal) if you want to poke around and see what I did.

All I did was add the modules from [this](https://github.com/nspsck/RM67162_Micropython_QSPI/tree/main) repo (same one linked above) as a user submodule, and built MicroPython normally otherwise (following the esp32 build instructions [here](https://github.com/micropython/micropython/tree/master/ports/esp32)). 

Then, I basically abused the RM67162 driver to work for my display. See my code [here](https://github.com/echo-lalia/beach-portal/blob/main/display/tft_config.py), where I basically have just repurposed the initialization routine from [here](https://github.com/moononournation/Arduino_GFX/blob/b3d190810b8e8c6d558b72176a247b2eb811241d/src/display/Arduino_NV3041A.h) without learning what any of it was for (but without doing that the colors looked bad).

I'm not sure how much of that would actually be useful for learning, because if I'm honest I don't have a great grasp on it myself. But I hope it can help someone 😅

**Reply by picpic020960** [reply]
Thank's for the answer.I will explored

**Reply by picpic020960** [reply]
OK works fine !
Now touch screen with gt911 to explore.
